homeassistant/components/mqtt/client.py

```diff
@@ -1141,8 +1141,8 @@ def _async_mqtt_on_callback(
         # see https://github.com/eclipse/paho.mqtt.python/issues/687
         # properties and reason codes are not used in Home Assistant
         future = self._async_get_mid_future(mid)
-        if future.done() and future.exception():
-            # Timed out
+        if future.done() and (future.cancelled() or future.exception()):
+            # Timed out or cancelled
             return
```

Comment:
Should  we not never set a result if the future is done?
So is it need to check if a future is cancelled or there is an exception, or are there more cases?

Reply:
Discussed on discord and decided to leave as-is